### DISCUSSION TITLE
Is multi-client supported?

### DISCUSSION THREAD

**Discussion by sabify** [maxcountryman/axum-login#224] [Ideas]
Hi,

Does `axum-login` support users who want to log in with multiple clients and keep their sessions active in case of having different hash tokens like OAuth tokens?

As I read the code and examples, it's possible through password hashing as it will be the same for all sessions in different clients, but what about other approaches like OAuth tokens?

**Comment by maxcountryman** [comment]
Are you asking if it's possible to log in with one method, say with a password, then log in with another method, say OAuth, and keep the same session?

As you point out, the auth hash would change, so you would need some common piece of secure data between the authentication methods that's a constant factor. One way of doing this would be to assign the hash internally, and store it with e.g. Argon2, alongside the user in your database.

Edit: I should point out that the goal of keeping the same session might be a little suspect; if both methods are associated with the same user, it would generally be better practice to not care about the session and allow the session manager to manage the session. If the concern is about what's in the session, then usually the correct thing to do there is handle loading data into the session within the application (again the idea is that the session is fungible).

**Comment by sabify** [comment]
I meant if we have only one method and it's OAuth login and a user tries to log in with multiple clients like Chrome on Desktop and Safari on iOS; with the current implementation (as you said), there should be only one internal hash token that will be used for all active sessions.

A hash change will unnecessarily invalidate all active sessions. How about having another option where we can map each user client's session to a different hash token?

**Reply by maxcountryman** [reply]
That's up to your implementation. You would likely want to generate a secure state that's managed by your backend and stored in your database. Then simply return this as your `auth_hash`.

**Reply by sabify** [reply]
First, thanks for you prompt attention.

Application logic on `auth_hash` may not work properly for multi-client users and the provided solution still has the same issue I described above.

I was thinking about how `from_session` (https://github.com/maxcountryman/axum-login/blob/48f045c66ead8be2895fbf9c062cec3e5b5d9f26/axum-login/src/session.rs#L144) works.

That method requests `session_auth_hash` by fetching the user with only the user ID. The only solution that comes to my mind for the current implementation is utilizing the generic `Id` associated type. Something like a pair of backend user ID and user-agent (or a random small hash) for `Id` can make each user's client distinguishable with their respective `auth_hash` (OAuth token).

For example:
```rust
#[derive(Clone, Serialize, Deserialize)]
pub struct UserId {
    id: i64,
    agent: String,
}
#[derive(Clone, Serialize, Deserialize, FromRow)]
pub struct User {
    id: UserId,
    pub access_token: String,
}

impl AuthUser for User {
    type Id = UserId;

    fn id(&self) -> Self::Id {
        self.id
    }

    fn session_auth_hash(&self) -> &[u8] {
        self.access_token.as_bytes()
    }
}
// ...
#[async_trait]
impl AuthnBackend for Backend {
    type User = User;
    type Credentials = Credentials;
    type Error = BackendError;

   // ...

    async fn get_user(&self, user_id: &UserId<Self>) -> Result<Option<Self::User>, Self::Error> {
        Ok(sqlx::query_as("select * from tokens where id = ? and agent = ?")
            .bind(user_id.id)
            .bind(user_id.agent)
            .fetch_optional(&self.db)
            .await
            .map_err(Self::Error::Sqlx)?)
    }
}
// ...
```

**Reply by maxcountryman** [reply]
> Application logic on auth_hash may not work properly for multi-client users and the provided solution still has the same issue I described above.

`auth_hash` already does what you want provided you manage the value.

One way of doing this is to create a unique hash for the user (like you would do for a password), this is inherently shared across all sessions and so it won't be invalidated when you log in on another user agent.

Taking a step back, this is the reason for this API: you're given full control over the semantics and can easily mold them to suit your use case.
